Support Consul back-end for Cluster Manager

BenjaminGill-Metaswitch opened the pull request:
In the orchestration team, we're planning to use Cluster Manager to manage a cluster of (containerised) Cassandra nodes, using a [Consul](https://www.consul.io/api/kv.html) back-end (instead of etcd) - see https://github.com/Metaswitch/dudamel/blob/master/docs/dm_pooling.md#cassandra

This PR adds support for a Consul back-end to the Cluster Manger as a replacement for etcd.

BenjaminGill-Metaswitch commented:
@murrayrogers - as discussed, please may you do the initial review?

murrayrogers commented on cluster_mgr_setup.py:
```diff
@@ -25,7 +25,8 @@
         "clearwater_etcd_shared",
-        "metaswitchcommon"],
+        "metaswitchcommon",
+        "python-consul==0.7.0"],
```
So will Clearwater need new dependencies, even though they're only used on DM?  Is Ellie OK with that?  (I don't imagine the impact is major - and fwiw third-party dependency stuff I'm sweeping up under the orchestration umbrella.)

murrayrogers commented on src/metaswitch/clearwater/cluster_manager/consul_synchronizer.py:
```diff
@@ -0,0 +1,174 @@
+#!/usr/bin/python
+
+# Copyright (C) Metaswitch Networks 2015
```
2015?

murrayrogers commented on src/metaswitch/clearwater/cluster_manager/main.py:
```diff
@@ -10,42 +10,49 @@
   -h --help                      Show this screen.
-  --mgmt-local-ip=IP             Management IP address
-  --sig-local-ip=IP              Signaling IP address
+  --mgmt-local-ip=IP             Management IP address (for the back-end)
```
what does "the back-end" mean?  Consul, or Cassandra?  If the former, is it "local"?

BenjaminGill-Metaswitch replied:
Consul/etcd

murrayrogers commented on src/metaswitch/clearwater/cluster_manager/main.py:
```diff
@@ -10,42 +10,49 @@
   --pidfile=FILE                 Pidfile to write [default: ./cluster-manager.pid]
   --cluster-manager-enabled=Y/N  Whether the cluster manager should start any threads [default: Yes]
+  --etcd                         Use the etcd back-end (the default)
+  --consul                       Use the Consul back-end
```
should this be --backend [etcd|consul]?  Might avoid worries about both being set?

BenjaminGill-Metaswitch replied:
From the usage: `[--etcd | --consul]`

murrayrogers replied:
OK

murrayrogers commented on src/metaswitch/clearwater/cluster_manager/main.py:
```diff
@@ -83,19 +90,25 @@ def main(args):
+    if arguments.get('--consul'):
+        backend = "consul"
+    else:
+        backend = "etcd"
```
hmm this doesn't obviously align with the arguments from higher up

BenjaminGill-Metaswitch replied:
`[--etcd | --consul]` forces only one to be present (and we default to etcd)

murrayrogers replied:
OK

murrayrogers commented on src/metaswitch/clearwater/cluster_manager/main.py:
```diff
@@ -107,14 +120,24 @@ def main(args):
     logging_config.configure_logging(log_level, log_dir, "cluster-manager", show_thread=True)
 
+    if foreground:
+        # In foreground mode, write logs to stdout as well
```
CWC change?  Ellie OK with it?

murrayrogers commented on src/metaswitch/clearwater/etcd_shared/test/mock_python_consul.py:
```diff
@@ -0,0 +1,196 @@
+# Copyright (C) Metaswitch Networks 2016
```
2016?

murrayrogers commented on src/metaswitch/clearwater/etcd_shared/common_consul_synchronizer.py:
```diff
@@ -0,0 +1,118 @@
+
+_log = logging.getLogger(__name__)
+
+class CommonConsulSynchronizer(CommonSynchronizer):
```
Not clear from comments how `ConsulSynchronizer` and `CommonConsulSynchronizer` differ - and their similar names possibly obfuscate the class hierarchy.  Are you able to add an extra sentence or two??

BenjaminGill-Metaswitch commented:
@eleanor-merry. As discussed a couple of weeks ago, we've now implemented the ability to choose a Consul back-end for the Clearwater cluster manager. Please may you review this (@murrayrogers has already done an internal review)?

The main changes are as follows:

* The back-end to use can now be chosen via the command line at run-time (defaults to `etcd`)
* There is now a consul synchronizer (common and for the cluster manager). This does the same thing as the etcd synchronizer, but for a Consul back-end
* Tests have been added to keep coverage at 100%
* Some dependencies have been added/removed to allow the cluster manager to be installed in a stand-alone fashion
* Print logs to stdout if running non-daemonised (makes containerisation easier)
* Added `cassandra_container_id` plugin parameter to support the new DDD Cassandra plugin for cluster manager

We've successfully been testing this against a live Consul and Cassandra.

BenjaminGill-Metaswitch commented:
@eleanor-merry - thanks for the useful discussion. You were nervous about the maintainability of this PR:

* Instead of having a ConsulSynchroniser (with lots of copied code), we should make the switch at a lower level - python-etcd and python-consul have similar APIs (modulo exceptions), so we should be able to have a `backend` class instead controlling the choice.
* It would be helpful to have some education about Consul prior to taking over maintenance. A lightning talk?
* Packaging: this PR strips out monit/etc. It might be a better idea to make 'cluster manager' its own package that gets pulled in (alongside monit/etc) by 'clearwater cluster manager'
* There are a number of errant references to 'etcd' in the Consul tests.

We've probably not got time to fix this in the next month, so I'm going to close and raise a story in our backlog to sort properly.

BenjaminGill-Metaswitch commented:
Issue raised as http://delphi/browse/ORCH-1689
